ouroboros-consensus/src/Ouroboros/Consensus/Protocol/PBFT/ChainState.hs

```diff
@@ -227,123 +316,154 @@ prune maxSize st
-                             Map.empty
-
-incrementKey :: Ord gk => gk -> Map gk Int -> Map gk Int
+incrementKey :: Ord gk => gk -> Map gk Word64 -> Map gk Word64
 incrementKey = Map.alter inc
   where
-    inc :: Maybe Int -> Maybe Int
+    inc :: Maybe Word64 -> Maybe Word64
     inc Nothing  = Just 1
     inc (Just n) = Just (n + 1)
 
-decrementKey :: Ord gk => gk -> Map gk Int -> Map gk Int
+decrementKey :: Ord gk => gk -> Map gk Word64 -> Map gk Word64
 decrementKey = Map.alter dec
   where
-    dec :: Maybe Int -> Maybe Int
+    dec :: Maybe Word64 -> Maybe Word64
     dec Nothing  = error "decrementKey: key does not exist"
     dec (Just 1) = Nothing
     dec (Just n) = Just (n - 1)
 
+-- | Internal: drop elements from the first list, keeping max size at most @k+n@
+prune :: SecurityParam
+      -> WindowSize
+      -> (StrictSeq a, StrictSeq a)
+      -> (StrictSeq a, StrictSeq a)
+prune (SecurityParam n) (WindowSize k) (xs, ys) =
+    (Seq.drop (fromIntegral toDrop) xs, ys)
+  where
+    totalSize, maxSize, toDrop :: Word64
+    totalSize = size xs + size ys
+    maxSize   = n + k
+    toDrop    = if totalSize > maxSize
+                  then totalSize - maxSize
+                  else 0
+
 {-------------------------------------------------------------------------------
-  Support for tests
+  Conversion
 -------------------------------------------------------------------------------}
 
--- | Construct PBFT chain state from map listing signed slots per key
-fromMap :: forall c. PBftCrypto c
-        => WithOrigin SlotNo -- ^ Anchor
-        -> Map (PBftVerKeyHash c) [SlotNo]
-        -> PBftChainState c
-fromMap anchor perKey = PBftChainState {
-      anchor  = anchor
-    , signers = perSlot
-    , counts  = fromSlots perSlot
-    }
+toList :: PBftChainState c -> (WithOrigin SlotNo, StrictSeq (PBftSigner c))
+toList PBftChainState{..} = (
+      case preAnchor of
+        Empty   -> Origin
+        _ :|> x -> At (pbftSignerSlotNo x)
+    , preWindow <> inWindow
+    )
+
+fromList :: PBftCrypto c
+         => SecurityParam
+         -> WindowSize
+         -> (WithOrigin SlotNo, StrictSeq (PBftSigner c))
+         -> PBftChainState c
+fromList k n (anchor, signers) = assertInvariant k n $ PBftChainState {..}
   where
-    perSlot :: StrictSeq (PBftSigner c)
-    perSlot = Seq.fromList
-            . sortOn pbftSignerSlotNo
-            . distrib
-            . Map.toList
-            $ perKey
+    inPreAnchor :: PBftSigner c -> Bool
+    inPreAnchor (PBftSigner slot _) = At slot <= anchor
 
-    distrib :: [(PBftVerKeyHash c, [SlotNo])] -> [PBftSigner c]
-    distrib = concatMap $ \(k, ss) -> map (\s -> PBftSigner s k) ss
+    (preAnchor, postAnchor) = Seq.spanl inPreAnchor signers
+    (preWindow, inWindow)   = Seq.splitAtEnd (fromIntegral n) signers
+    counts                  = computeCounts inWindow
 
 {-------------------------------------------------------------------------------
   Serialization
```

Comment:
I claim that if we start with a list containing too few keys (specifically `n` keys, not `n + k` keys, like in existing deployments), after `k` number of `append`s the invariant will be restored. It might be good to test this. 

Rewinding it a tad trickier, but I think it will still be okay, at least for rewinds for strictly `< k`; all that would happen is we might end up with a shorter window size. (We could test this also, potentially.) This is a calculated risk. If we don't want to take this risk, we should change the serialization format, forcing all nodes to replay all blocks from genesis, rebuilding a ledger snapshot.

/cc @dcoutts